File: clients/python/examples/basic_usage.py
#!/usr/bin/env python3
"""Basic usage examples for the Overture Geocoder client.

Run with: python examples/basic_usage.py
"""

from overture_geocoder import OvertureGeocoder, geocode


def main():
    # =========================================================================
    # Quick functions (use default configuration)
    # =========================================================================

    print("=== Quick geocode function ===")
    quick_results = geocode("123 Main St, Boston, MA")
    print(f"Found: {len(quick_results)} results")
    if quick_results:
        print(f"First result: {quick_results[0].display_name}")

    # =========================================================================
    # Using the client class
    # =========================================================================

    print("\n=== OvertureGeocoder client ===")

    # Create a client with custom configuration
    with OvertureGeocoder(
        base_url="http://localhost:8787",  # Default
        timeout=10.0,  # 10 seconds
    ) as client:

        # Basic search
        print("\n--- Basic search ---")
        results = client.search("Boston City Hall")
        print(f"Search results: {len(results)}")
        for result in results[:3]:
            print(f"  - {result.display_name}")
            print(f"    Lat: {result.lat}, Lon: {result.lon}")
            print(f"    GERS ID: {result.gers_id}")

        # =====================================================================
        # Search options
        # =====================================================================

        print("\n--- Search with options ---")

        # Limit results
        limited = client.search("Main St", limit=5)
        print(f"Limited to 5 results: {len(limited)}")

        # Include address details
        with_address = client.search("123 Main St", addressdetails=True, limit=1)
        if with_address and with_address[0].address:
            addr = with_address[0].address
            print("Address breakdown:")
            print(f"  House number: {addr.house_number}")
            print(f"  Road: {addr.road}")
            print(f"  City: {addr.city}")
            print(f"  State: {addr.state}")
            print(f"  Postcode: {addr.postcode}")

        # Search within a bounding box
        in_box = client.search(
            "coffee",
            viewbox=(-71.1, 42.3, -71.0, 42.4),  # (lon1, lat1, lon2, lat2)
            bounded=True,
            limit=3,
        )
        print(f"Results in bounding box: {len(in_box)}")

        # =====================================================================
        # GeoJSON format
        # =====================================================================

        print("\n--- GeoJSON format ---")

        geojson = client.search_geojson("Boston Public Library")
        print(f"GeoJSON type: {geojson['type']}")
        print(f"Features: {len(geojson['features'])}")
        if geojson["features"]:
            feature = geojson["features"][0]
            print("First feature:")
            print(f"  ID: {feature['id']}")
            print(f"  Coordinates: {feature['geometry']['coordinates']}")

        # =====================================================================
        # Get full geometry by GERS ID (fetches directly from Overture S3)
        # =====================================================================

        print("\n--- Get full geometry by GERS ID ---")

        if results:
            gers_id = results[0].gers_id
            print(f"Fetching geometry for GERS ID: {gers_id}")

            # Get full geometry (requires overturemaps package)
            try:
                feature = client.get_geometry(gers_id)
                if feature:
                    print(f"Geometry type: {feature['geometry']['type']}")
                    print(f"Properties: {list(feature['properties'].keys())}")
            except ImportError as e:
                print(f"Skipping geometry fetch: {e}")

    print("\n=== Done ===")


if __name__ == "__main__":
    main()

File: clients/js/examples/basic-usage.ts
/**
 * Basic usage examples for the Overture Geocoder client.
 *
 * Run with: npx tsx examples/basic-usage.ts
 */

import { OvertureGeocoder, geocode, lookup } from "../src/index";

async function main() {
  // ==========================================================================
  // Quick functions (use default configuration)
  // ==========================================================================

  console.log("=== Quick geocode function ===");
  const quickResults = await geocode("123 Main St, Boston, MA");
  console.log("Found:", quickResults.length, "results");
  if (quickResults.length > 0) {
    console.log("First result:", quickResults[0].display_name);
  }

  // ==========================================================================
  // Using the client class
  // ==========================================================================

  console.log("\n=== OvertureGeocoder client ===");

  // Create a client with custom configuration
  const client = new OvertureGeocoder({
    baseUrl: "http://localhost:8787", // Default
    timeout: 10000, // 10 seconds
  });

  // Basic search
  console.log("\n--- Basic search ---");
  const results = await client.search("Boston City Hall");
  console.log("Search results:", results.length);
  for (const result of results.slice(0, 3)) {
    console.log(`  - ${result.display_name}`);
    console.log(`    Lat: ${result.lat}, Lon: ${result.lon}`);
    console.log(`    GERS ID: ${result.gers_id}`);
  }

  // ==========================================================================
  // Search options
  // ==========================================================================

  console.log("\n--- Search with options ---");

  // Limit results
  const limited = await client.search("Main St", { limit: 5 });
  console.log("Limited to 5 results:", limited.length);

  // Include address details
  const withAddress = await client.search("123 Main St", {
    addressdetails: true,
    limit: 1,
  });
  if (withAddress.length > 0 && withAddress[0].address) {
    console.log("Address breakdown:");
    console.log("  House number:", withAddress[0].address.house_number);
    console.log("  Road:", withAddress[0].address.road);
    console.log("  City:", withAddress[0].address.city);
    console.log("  State:", withAddress[0].address.state);
    console.log("  Postcode:", withAddress[0].address.postcode);
  }

  // Search within a bounding box
  const inBox = await client.search("coffee", {
    viewbox: [-71.1, 42.3, -71.0, 42.4], // [lon1, lat1, lon2, lat2]
    bounded: true,
    limit: 3,
  });
  console.log("Results in bounding box:", inBox.length);

  // ==========================================================================
  // GeoJSON format
  // ==========================================================================

  console.log("\n--- GeoJSON format ---");

  const geojson = await client.searchGeoJSON("Boston Public Library");
  console.log("GeoJSON type:", geojson.type);
  console.log("Features:", geojson.features.length);
  if (geojson.features.length > 0) {
    const feature = geojson.features[0];
    console.log("First feature:");
    console.log("  ID:", feature.id);
    console.log("  Coordinates:", feature.geometry.coordinates);
  }

  // ==========================================================================
  // Lookup by GERS ID
  // ==========================================================================

  console.log("\n--- Lookup by GERS ID ---");

  if (results.length > 0) {
    const gersId = results[0].gers_id;
    console.log("Looking up GERS ID:", gersId);

    // Single lookup
    const lookupResult = await client.lookup(gersId);
    console.log("Lookup result:", lookupResult[0]?.display_name);

    // Multiple lookups
    if (results.length > 1) {
      const ids = results.slice(0, 3).map((r) => r.gers_id);
      const multiLookup = await client.lookup(ids);
      console.log("Multi-lookup results:", multiLookup.length);
    }

    // Get geometry
    const geometry = await client.getGeometry(gersId);
    if (geometry) {
      console.log("Geometry type:", geometry.geometry.type);
      console.log("Coordinates:", geometry.geometry.coordinates);
    }
  }

  console.log("\n=== Done ===");
}

main().catch(console.error);
